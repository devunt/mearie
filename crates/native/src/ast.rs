use crate::span::*;
use bumpalo::collections::Vec;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Name<'a>(&'a str);

impl<'a> Name<'a> {
    #[inline]
    pub const fn new(value: &'a str) -> Self {
        Self(value)
    }

    #[inline]
    pub const fn as_str(self) -> &'a str {
        self.0
    }
}

impl<'a> From<&'a str> for Name<'a> {
    #[inline]
    fn from(s: &'a str) -> Self {
        Self::new(s)
    }
}

impl<'a> std::ops::Deref for Name<'a> {
    type Target = str;

    #[inline]
    fn deref(&self) -> &Self::Target {
        self.0
    }
}

impl<'a> std::fmt::Display for Name<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.0)
    }
}

impl<'a> PartialEq<str> for Name<'a> {
    #[inline]
    fn eq(&self, other: &str) -> bool {
        self.0 == other
    }
}

impl<'a> PartialEq<&str> for Name<'a> {
    #[inline]
    fn eq(&self, other: &&str) -> bool {
        self.0 == *other
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Value<'a> {
    Variable(Name<'a>),
    Int(&'a str),
    Float(&'a str),
    String(&'a str),
    Boolean(bool),
    Null,
    Enum(Name<'a>),
    List(Vec<'a, Value<'a>>),
    Object(Vec<'a, ObjectField<'a>>),
}

#[allow(dead_code)]
impl<'a> Value<'a> {
    #[inline]
    pub fn as_string(&self) -> Option<&'a str> {
        match self {
            Value::String(s) => Some(s),
            _ => None,
        }
    }

    #[inline]
    pub fn as_int(&self) -> Option<Result<i64, std::num::ParseIntError>> {
        match self {
            Value::Int(s) => Some(s.parse()),
            _ => None,
        }
    }

    #[inline]
    pub fn as_float(&self) -> Option<Result<f64, std::num::ParseFloatError>> {
        match self {
            Value::Float(s) => Some(s.parse()),
            _ => None,
        }
    }

    #[inline]
    pub fn as_bool(&self) -> Option<bool> {
        match self {
            Value::Boolean(b) => Some(*b),
            _ => None,
        }
    }

    #[inline]
    pub fn is_null(&self) -> bool {
        matches!(self, Value::Null)
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct ObjectField<'a> {
    pub name: Name<'a>,
    pub value: Value<'a>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Type<'a> {
    Named(NamedType<'a>),
    List(&'a Type<'a>),
    NonNull(&'a NonNullType<'a>),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct NamedType<'a> {
    pub name: &'a str,
}

impl<'a> NamedType<'a> {
    #[inline]
    pub const fn new(name: &'a str) -> Self {
        Self { name }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum NonNullType<'a> {
    Named(NamedType<'a>),
    List(&'a Type<'a>),
}

#[derive(Debug, Clone, PartialEq)]
pub struct Directive<'a> {
    pub span: Span,
    pub name: Name<'a>,
    pub arguments: Vec<'a, Argument<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Argument<'a> {
    pub span: Span,
    pub name: Name<'a>,
    pub value: Value<'a>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Description<'a> {
    pub value: &'a str,
}

#[derive(Debug, Clone, PartialEq)]
pub struct OperationDefinition<'a> {
    pub span: Span,
    pub operation_type: OperationType,
    pub name: Option<Name<'a>>,
    pub variable_definitions: Vec<'a, VariableDefinition<'a>>,
    pub directives: Vec<'a, Directive<'a>>,
    pub selection_set: SelectionSet<'a>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum OperationType {
    Query,
    Mutation,
    Subscription,
}

#[derive(Debug, Clone, PartialEq)]
pub struct VariableDefinition<'a> {
    pub span: Span,
    pub variable: Name<'a>,
    pub typ: Type<'a>,
    pub default_value: Option<Value<'a>>,
    pub directives: Vec<'a, Directive<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct SelectionSet<'a> {
    pub selections: Vec<'a, Selection<'a>>,
}

impl<'a> SelectionSet<'a> {
    #[inline]
    pub fn is_empty(&self) -> bool {
        self.selections.is_empty()
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Selection<'a> {
    Field(Field<'a>),
    FragmentSpread(FragmentSpread<'a>),
    InlineFragment(InlineFragment<'a>),
}

#[derive(Debug, Clone, PartialEq)]
pub struct Field<'a> {
    pub span: Span,
    pub alias: Option<Name<'a>>,
    pub name: Name<'a>,
    pub arguments: Vec<'a, Argument<'a>>,
    pub directives: Vec<'a, Directive<'a>>,
    pub selection_set: SelectionSet<'a>,
}

impl<'a> Field<'a> {
    #[inline]
    pub fn alias_or_name(&self) -> Name<'a> {
        self.alias.unwrap_or(self.name)
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct FragmentSpread<'a> {
    pub span: Span,
    pub fragment_name: Name<'a>,
    pub directives: Vec<'a, Directive<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct InlineFragment<'a> {
    pub span: Span,
    pub type_condition: Option<Name<'a>>,
    pub directives: Vec<'a, Directive<'a>>,
    pub selection_set: SelectionSet<'a>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct FragmentDefinition<'a> {
    pub span: Span,
    pub name: Name<'a>,
    pub type_condition: Name<'a>,
    pub directives: Vec<'a, Directive<'a>>,
    pub selection_set: SelectionSet<'a>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct SchemaDefinition<'a> {
    pub description: Option<Description<'a>>,
    pub directives: Vec<'a, Directive<'a>>,
    pub query: Option<Name<'a>>,
    pub mutation: Option<Name<'a>>,
    pub subscription: Option<Name<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct SchemaExtension<'a> {
    pub directives: Vec<'a, Directive<'a>>,
    pub query: Option<Name<'a>>,
    pub mutation: Option<Name<'a>>,
    pub subscription: Option<Name<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum TypeDefinition<'a> {
    Scalar(ScalarTypeDefinition<'a>),
    Object(ObjectTypeDefinition<'a>),
    Interface(InterfaceTypeDefinition<'a>),
    Union(UnionTypeDefinition<'a>),
    Enum(EnumTypeDefinition<'a>),
    InputObject(InputObjectTypeDefinition<'a>),
}

#[derive(Debug, Clone, PartialEq)]
pub enum TypeExtension<'a> {
    Scalar(ScalarTypeExtension<'a>),
    Object(ObjectTypeExtension<'a>),
    Interface(InterfaceTypeExtension<'a>),
    Union(UnionTypeExtension<'a>),
    Enum(EnumTypeExtension<'a>),
    InputObject(InputObjectTypeExtension<'a>),
}

#[derive(Debug, Clone, PartialEq)]
pub struct ScalarTypeDefinition<'a> {
    pub description: Option<Description<'a>>,
    pub name: Name<'a>,
    pub directives: Vec<'a, Directive<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ScalarTypeExtension<'a> {
    pub name: Name<'a>,
    pub directives: Vec<'a, Directive<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ObjectTypeDefinition<'a> {
    pub description: Option<Description<'a>>,
    pub name: Name<'a>,
    pub implements: Vec<'a, Name<'a>>,
    pub directives: Vec<'a, Directive<'a>>,
    pub fields: Vec<'a, FieldDefinition<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ObjectTypeExtension<'a> {
    pub name: Name<'a>,
    pub implements: Vec<'a, Name<'a>>,
    pub directives: Vec<'a, Directive<'a>>,
    pub fields: Vec<'a, FieldDefinition<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct FieldDefinition<'a> {
    pub description: Option<Description<'a>>,
    pub name: Name<'a>,
    pub arguments: Vec<'a, InputValueDefinition<'a>>,
    pub typ: Type<'a>,
    pub directives: Vec<'a, Directive<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct InputValueDefinition<'a> {
    pub description: Option<Description<'a>>,
    pub name: Name<'a>,
    pub typ: Type<'a>,
    pub default_value: Option<Value<'a>>,
    pub directives: Vec<'a, Directive<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct InterfaceTypeDefinition<'a> {
    pub description: Option<Description<'a>>,
    pub name: Name<'a>,
    pub implements: Vec<'a, Name<'a>>,
    pub directives: Vec<'a, Directive<'a>>,
    pub fields: Vec<'a, FieldDefinition<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct InterfaceTypeExtension<'a> {
    pub name: Name<'a>,
    pub implements: Vec<'a, Name<'a>>,
    pub directives: Vec<'a, Directive<'a>>,
    pub fields: Vec<'a, FieldDefinition<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct UnionTypeDefinition<'a> {
    pub description: Option<Description<'a>>,
    pub name: Name<'a>,
    pub directives: Vec<'a, Directive<'a>>,
    pub members: Vec<'a, Name<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct UnionTypeExtension<'a> {
    pub name: Name<'a>,
    pub directives: Vec<'a, Directive<'a>>,
    pub members: Vec<'a, Name<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct EnumTypeDefinition<'a> {
    pub description: Option<Description<'a>>,
    pub name: Name<'a>,
    pub directives: Vec<'a, Directive<'a>>,
    pub values: Vec<'a, EnumValueDefinition<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct EnumTypeExtension<'a> {
    pub name: Name<'a>,
    pub directives: Vec<'a, Directive<'a>>,
    pub values: Vec<'a, EnumValueDefinition<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct EnumValueDefinition<'a> {
    pub description: Option<Description<'a>>,
    pub value: Name<'a>,
    pub directives: Vec<'a, Directive<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct InputObjectTypeDefinition<'a> {
    pub description: Option<Description<'a>>,
    pub name: Name<'a>,
    pub directives: Vec<'a, Directive<'a>>,
    pub fields: Vec<'a, InputValueDefinition<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct InputObjectTypeExtension<'a> {
    pub name: Name<'a>,
    pub directives: Vec<'a, Directive<'a>>,
    pub fields: Vec<'a, InputValueDefinition<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct DirectiveDefinition<'a> {
    pub description: Option<Description<'a>>,
    pub name: Name<'a>,
    pub arguments: Vec<'a, InputValueDefinition<'a>>,
    pub repeatable: bool,
    pub locations: Vec<'a, DirectiveLocation>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum DirectiveLocation {
    Query,
    Mutation,
    Subscription,
    Field,
    FragmentDefinition,
    FragmentSpread,
    InlineFragment,
    VariableDefinition,
    Schema,
    Scalar,
    Object,
    FieldDefinition,
    ArgumentDefinition,
    Interface,
    Union,
    Enum,
    EnumValue,
    InputObject,
    InputFieldDefinition,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Document<'a> {
    pub source: &'a Source<'a>,
    pub definitions: Vec<'a, Definition<'a>>,
}

impl<'a> Document<'a> {
    #[inline]
    pub fn is_empty(&self) -> bool {
        self.definitions.is_empty()
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Definition<'a> {
    Executable(ExecutableDefinition<'a>),
    TypeSystem(TypeSystemDefinition<'a>),
    TypeSystemExtension(TypeSystemExtension<'a>),
}

#[derive(Debug, Clone, PartialEq)]
pub enum ExecutableDefinition<'a> {
    Operation(OperationDefinition<'a>),
    Fragment(FragmentDefinition<'a>),
}

#[derive(Debug, Clone, PartialEq)]
pub enum TypeSystemDefinition<'a> {
    Schema(SchemaDefinition<'a>),
    Type(TypeDefinition<'a>),
    Directive(DirectiveDefinition<'a>),
}

#[derive(Debug, Clone, PartialEq)]
pub enum TypeSystemExtension<'a> {
    Schema(SchemaExtension<'a>),
    Type(TypeExtension<'a>),
}
