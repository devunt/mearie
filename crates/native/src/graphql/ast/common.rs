use std::fmt;
use std::ops::Deref;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Name<'a>(&'a str);

impl<'a> Name<'a> {
    /// Creates a new Name from a string slice.
    #[inline]
    pub const fn new(value: &'a str) -> Self {
        Self(value)
    }

    /// Returns the underlying string slice.
    #[inline]
    pub const fn as_str(self) -> &'a str {
        self.0
    }
}

impl<'a> From<&'a str> for Name<'a> {
    #[inline]
    fn from(s: &'a str) -> Self {
        Self::new(s)
    }
}

impl<'a> Deref for Name<'a> {
    type Target = str;

    #[inline]
    fn deref(&self) -> &Self::Target {
        self.0
    }
}

impl<'a> fmt::Display for Name<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.0)
    }
}

impl<'a> PartialEq<str> for Name<'a> {
    #[inline]
    fn eq(&self, other: &str) -> bool {
        self.0 == other
    }
}

impl<'a> PartialEq<&str> for Name<'a> {
    #[inline]
    fn eq(&self, other: &&str) -> bool {
        self.0 == *other
    }
}

/// Represents a GraphQL description string that can be attached to various definitions.
#[derive(Debug, Clone, PartialEq)]
pub struct Description<'a> {
    pub value: &'a str,
}
