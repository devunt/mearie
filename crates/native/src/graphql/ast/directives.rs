use super::types::{ArgumentName, DirectiveName};
use super::values::Value;
use crate::error::location::Span;
use bumpalo::collections::Vec;

/// Represents a GraphQL directive with its name and arguments.
#[derive(Debug, Clone, PartialEq)]
pub struct Directive<'a> {
    pub span: Span,
    pub name: DirectiveName<'a>,
    pub arguments: Vec<'a, Argument<'a>>,
}

impl<'a> Directive<'a> {
    /// Gets the value of an argument by name.
    pub fn get_argument(&self, name: &str) -> Option<&Value<'a>> {
        self.arguments.iter().find(|arg| arg.name == name).map(|arg| &arg.value)
    }

    /// Returns true if this directive has an argument with the given name.
    #[inline]
    pub fn has_argument(&self, name: &str) -> bool {
        self.arguments.iter().any(|arg| arg.name == name)
    }
}

/// Represents an argument in a directive or field.
#[derive(Debug, Clone, PartialEq)]
pub struct Argument<'a> {
    pub span: Span,
    pub name: ArgumentName<'a>,
    pub value: Value<'a>,
}

/// Represents the locations where a directive can be applied in a GraphQL schema.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum DirectiveLocation {
    Query,
    Mutation,
    Subscription,
    Field,
    FragmentDefinition,
    FragmentSpread,
    InlineFragment,
    VariableDefinition,
    Schema,
    Scalar,
    Object,
    FieldDefinition,
    ArgumentDefinition,
    Interface,
    Union,
    Enum,
    EnumValue,
    InputObject,
    InputFieldDefinition,
}
