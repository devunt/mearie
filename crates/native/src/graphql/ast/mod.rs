pub mod common;
pub mod directives;
pub mod document;
pub mod fragments;
pub mod operations;
pub mod types;
pub mod values;

pub use common::*;
pub use directives::*;
pub use document::*;
pub use fragments::*;
pub use operations::*;
pub use types::*;
pub use values::*;
