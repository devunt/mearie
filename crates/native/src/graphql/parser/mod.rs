pub mod lexer;
#[allow(clippy::module_inception)]
pub mod parser;
pub mod state;

pub use parser::Parser;
pub use state::{Complete, Failed, Parsing, Ready, Uninitialized};
