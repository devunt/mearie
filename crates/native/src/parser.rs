pub mod context;
pub mod lexer;
pub mod parse;
pub mod printer;

pub use context::GraphQLContext;
pub use parse::ParseNode;
