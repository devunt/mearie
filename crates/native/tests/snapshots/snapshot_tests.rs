use insta::assert_debug_snapshot;
use mearie_native::ast::Document;
use mearie_native::parser::{GraphQLContext, ParseNode};
use mearie_native::span::Source;

#[test]
fn test_snapshot_simple_query() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "query { user { id name } }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_query_with_variables() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "query GetUser($id: ID!) { user(id: $id) { id name } }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_mutation() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "mutation CreateUser { createUser(name: \"John\") { id } }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_fragment() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "fragment UserFields on User { id name email }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_inline_fragment() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "query { search { ... on User { name } } }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_type_definition() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "type User { id: ID! name: String! email: String }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_interface_definition() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "interface Node { id: ID! createdAt: String }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_union_definition() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "union SearchResult = User | Post | Comment",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_enum_definition() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "enum Status { ACTIVE INACTIVE PENDING }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_input_definition() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "input UserInput { name: String! email: String! age: Int }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_scalar_definition() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "scalar DateTime",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_directive_definition() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "directive @auth(requires: Role!) on FIELD_DEFINITION",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_schema_definition() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "schema { query: Query mutation: Mutation }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_extend_type() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "extend type User { avatar: String }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_query_with_directives() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "query { user @skip(if: true) { id } }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_query_with_aliases() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "query { userId: user { userName: name } }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_nested_selection_sets() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "query { user { posts { comments { author { name } } } } }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_query_with_default_values() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "query($limit: Int = 10) { users(first: $limit) { id } }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_subscription() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "subscription { messageAdded { id content } }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_type_with_implements() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "type User implements Node { id: ID! }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_type_with_multiple_interfaces() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "type User implements Node & Timestamped { id: ID! }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_field_with_arguments() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "type Query { users(first: Int, offset: Int): [User] }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_list_type() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "type User { tags: [String] }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_non_null_type() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "type User { id: ID! name: String! }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_non_null_list() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "type User { tags: [String!]! }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_type_with_description() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: r#""A user in the system" type User { id: ID! }"#,
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_field_with_description() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: r#"type User { "User ID" id: ID! }"#,
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_enum_with_description() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: r#""User status" enum Status { ACTIVE INACTIVE }"#,
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_input_with_default_values() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "input UserInput { name: String! age: Int = 18 }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_repeatable_directive() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "directive @tag(name: String!) repeatable on FIELD",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_multiple_operations() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: r#"
        query GetUser { user { id } }
        query GetPosts { posts { id } }
    "#,
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_query_with_fragment_spread() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: r#"
        query { user { ...UserFields } }
        fragment UserFields on User { id name }
    "#,
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_nested_fragments() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: r#"
        query { user { ...UserWithPosts } }
        fragment UserWithPosts on User { id ...UserBasic }
        fragment UserBasic on User { name email }
    "#,
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_complex_query() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: r#"
        query GetUserData($id: ID!, $includeEmail: Boolean = false) {
            user(id: $id) {
                id
                name
                email @include(if: $includeEmail)
                posts(first: 10) {
                    edges {
                        node {
                            title
                        }
                    }
                }
            }
        }
    "#,
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_query_with_list_argument() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "query { users(ids: [1, 2, 3]) { id } }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_query_with_object_argument() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "query { createUser(input: { name: \"John\", age: 30 }) { id } }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_extend_interface() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "extend interface Node { createdAt: String }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_schema_with_all_operations() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "schema { query: Query mutation: Mutation subscription: Subscription }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_interface_with_fields() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "interface Node { id: ID! createdAt: String! }",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}

#[test]
fn test_snapshot_union_with_multiple_types() {
    let ctx = GraphQLContext::new();
    let source = Source {
        code: "union SearchResult = User | Post | Comment",
        file_path: "test.graphql",
        start_line: 0,
    };
    let result = Document::parse(&ctx, &source).unwrap();
    assert_debug_snapshot!(result);
}
