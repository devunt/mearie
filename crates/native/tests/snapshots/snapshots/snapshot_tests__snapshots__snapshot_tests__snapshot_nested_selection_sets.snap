---
source: native/tests/snapshots/snapshot_tests.rs
expression: result
---
Document {
    source: Source {
        code: "query { user { posts { comments { author { name } } } } }",
        file_path: "test.graphql",
        start_line: 0,
    },
    definitions: [
        Executable(
            Operation(
                OperationDefinition {
                    span: Span {
                        start: 0,
                        end: 57,
                    },
                    operation_type: Query,
                    name: None,
                    variable_definitions: [],
                    directives: [],
                    selection_set: SelectionSet {
                        selections: [
                            Field(
                                Field {
                                    span: Span {
                                        start: 8,
                                        end: 55,
                                    },
                                    alias: None,
                                    name: Name(
                                        "user",
                                    ),
                                    arguments: [],
                                    directives: [],
                                    selection_set: SelectionSet {
                                        selections: [
                                            Field(
                                                Field {
                                                    span: Span {
                                                        start: 15,
                                                        end: 53,
                                                    },
                                                    alias: None,
                                                    name: Name(
                                                        "posts",
                                                    ),
                                                    arguments: [],
                                                    directives: [],
                                                    selection_set: SelectionSet {
                                                        selections: [
                                                            Field(
                                                                Field {
                                                                    span: Span {
                                                                        start: 23,
                                                                        end: 51,
                                                                    },
                                                                    alias: None,
                                                                    name: Name(
                                                                        "comments",
                                                                    ),
                                                                    arguments: [],
                                                                    directives: [],
                                                                    selection_set: SelectionSet {
                                                                        selections: [
                                                                            Field(
                                                                                Field {
                                                                                    span: Span {
                                                                                        start: 34,
                                                                                        end: 49,
                                                                                    },
                                                                                    alias: None,
                                                                                    name: Name(
                                                                                        "author",
                                                                                    ),
                                                                                    arguments: [],
                                                                                    directives: [],
                                                                                    selection_set: SelectionSet {
                                                                                        selections: [
                                                                                            Field(
                                                                                                Field {
                                                                                                    span: Span {
                                                                                                        start: 43,
                                                                                                        end: 49,
                                                                                                    },
                                                                                                    alias: None,
                                                                                                    name: Name(
                                                                                                        "name",
                                                                                                    ),
                                                                                                    arguments: [],
                                                                                                    directives: [],
                                                                                                    selection_set: SelectionSet {
                                                                                                        selections: [],
                                                                                                    },
                                                                                                },
                                                                                            ),
                                                                                        ],
                                                                                    },
                                                                                },
                                                                            ),
                                                                        ],
                                                                    },
                                                                },
                                                            ),
                                                        ],
                                                    },
                                                },
                                            ),
                                        ],
                                    },
                                },
                            ),
                        ],
                    },
                },
            ),
        ),
    ],
}
